### PR TITLE
New Entities TemplateImport and TemplateExport

### PR DESCRIPTION
**Pull request by jyejare**
$subject

### CODE REVIEW

**Comment by jyejare** [comment]
@pondrejk , @abalakh , Need your eyes on the PR.

File: nailgun/entities.py [hunk 1]
```diff
@@ -6579,6 +6607,81 @@ def read(self, entity=None, attrs=None, ignore=None, params=None):
...
+    """A representation of a Template Import entity."""
+
+    def __init__(self, server_config=None, **kwargs):
+        self._fields = {
+            'associate': entity_fields.StringField(),
+            'branch': entity_fields.StringField(),
+            'dirname': entity_fields.StringField(),
+            'filter': entity_fields.StringField(),
+            'force': entity_fields.StringField(),
+            'location_id': entity_fields.IntegerField(),
```
Comment (abalakh) [inline nailgun/entities.py]: Should be `'location': entity_fields.OneToOneField(Location)` instead

File: nailgun/entities.py [hunk 2]
```diff
@@ -6579,6 +6607,81 @@ def read(self, entity=None, attrs=None, ignore=None, params=None):
...
+            'lock': entity_fields.BooleanField(),
+            'negate': entity_fields.BooleanField(),
+            'organization_id': entity_fields.IntegerField(),
```
Comment (abalakh) [inline nailgun/entities.py]: `'organization': entity_fields.OneToOneField(Organization),`

File: nailgun/entities.py [hunk 3]
```diff
@@ -6579,6 +6607,81 @@ def read(self, entity=None, attrs=None, ignore=None, params=None):
...
+            'prefix': entity_fields.StringField(),
+            'repo': entity_fields.StringField(),
+            'templates': entity_fields.ListField(),
```
Comment (abalakh) [inline nailgun/entities.py]: What's inside that list? Maybe it should be replaced with OneToManyField(SomeEntity)

File: nailgun/entities.py [hunk 4]
```diff
@@ -6579,6 +6607,81 @@ def read(self, entity=None, attrs=None, ignore=None, params=None):
...
+
+class TemplateExport(Entity, EntityCreateMixin, EntityReadMixin):
+    """A representation of a Template Export entity."""
+
+    def __init__(self, server_config=None, **kwargs):
+        self._fields = {
+            'branch': entity_fields.StringField(),
+            'dirname': entity_fields.StringField(),
+            'filter': entity_fields.StringField(),
+            'location_ids': entity_fields.IntegerField(),
```
Comment (abalakh) [inline nailgun/entities.py]: Same here, replace with `entity_fields.OneToManyField(Location)`

File: nailgun/entities.py (hunk 3 above)
Comment (jyejare) [reply nailgun/entities.py]: Yeah, Its output that import/export of templates returns ```[{'imported': yes/no, name:'some1'}, {{'imported': yes/no, name:'some2'}}]```(also many other attributes) for which we may need to create new entity class. Actually, I wanted to create that but I see nailgun doesn't have such things for any entity so just skipped my idea and stick with the list. 

Any thoughts ?

Comment (abalakh) [reply nailgun/entities.py]: Ok, not a blocker then. Pls just add some `# todo` comment so we won't forget

File: nailgun/entities.py [hunk 5]
```diff
@@ -3819,6 +3819,34 @@ def search_normalize(self, results):
         return super(Interface, self).search_normalize(results)
 
 
+class JobTemplate(
+        Entity,
+        EntityReadMixin,
+        EntitySearchMixin):
+    """A representation of a Provisioning Template entity."""
```
Comment (manuelbonk) [inline nailgun/entities.py]: Don't you mean `Job Template entity`?

File: nailgun/entities.py [hunk 6]
```diff
@@ -3819,6 +3819,34 @@ def search_normalize(self, results):
...
+
+    def __init__(self, server_config=None, **kwargs):
+        self._fields = {
```
Comment (manuelbonk) [inline nailgun/entities.py]: I think there are some fields missing like template inputs, audit comment, effective user, and description format.

Comment (jyejare) [reply nailgun/entities.py]: @manuelbonk , Actually I have created that template just for the view job template purpose for Foreman Templates future. So the minimal params are populated. I can remove the entire entity in favor of ur PR https://github.com/SatelliteQE/nailgun/pull/501.

File: nailgun/entities.py (hunk 5 above)
Comment (jyejare) [reply nailgun/entities.py]: yup, JobTemplate.

**Comment by jyejare** [comment]
This PR is depend on #501

**Comment by jyejare** [comment]
As per @abalakh  comment adding Create() console output for TemplateImport and TemplateExport entities:
PS: The read() cannot be called on this entities directly, it is just used when create returns data.

```
TemplateImport Entity:

entities.TemplateImport(repo='https://github.com/jyejare/foreman_templates.git', organization=1, 
   ...: filter='robottelo').create()
Out[4]: nailgun.entities.TemplateImport(templates=[{u'validation_errors': {}, u'exception': None, u'imported': False, u'name': u'Test Data', u'changed': False, u'additional_errors': u"Skipping, 'name' filtered out based on 'filter' and 'negate' settings", u'file': u'metadata1.erb', u'type': None, u'id': None}, {u'validation_errors': {}, u'exception': None, u'imported': False, u'name': u'Model-based Test', u'changed': False, u'additional_errors': u"Skipping, 'name' filtered out based on 'filter' and 'negate' settings", u'file': u'metadata2.erb', u'type': None, u'id': None}, {u'validation_errors': {}, u'exception': None, u'imported': False, u'name': u'Test Ptable', u'changed': False, u'additional_errors': u"Skipping, 'name' filtered out based on 'filter' and 'negate' settings", u'file': u'ptable1.erb', u'type': None, u'id': None}, {u'validation_errors': {}, u'exception': None, u'imported': False, u'name': u'Test Snippet', u'changed': False, u'additional_errors': u"Skipping, 'name' filtered out based on 'filter' and 'negate' settings", u'file': u'snippet1.erb', u'type': None, u'id': None}, {u'validation_errors': {}, u'exception': None, u'imported': False, u'name': u'Test Data', u'changed': False, u'additional_errors': u"Skipping, 'name' filtered out based on 'filter' and 'negate' settings", u'file': u'metadata1.erb', u'type': None, u'id': None}, {u'validation_errors': {}, u'exception': None, u'imported': False, u'name': u'Test Ptable', u'changed': False, u'additional_errors': u"Skipping, 'name' filtered out based on 'filter' and 'negate' settings", u'file': u'ptable1.erb', u'type': None, u'id': None}, {u'validation_errors': {}, u'exception': None, u'imported': False, u'name': u'Test Snippet', u'changed': False, u'additional_errors': u"Skipping, 'name' filtered out based on 'filter' and 'negate' settings", u'file': u'snippet1.erb', u'type': None, u'id': None}, {u'validation_errors': {}, u'exception': None, u'imported': False, u'name': u'Test Data', u'changed': False, u'additional_errors': u"Skipping, 'name' filtered out based on 'filter' and 'negate' settings", u'file': u'metadata1.erb', u'type': None, u'id': None}, {u'validation_errors': {}, u'exception': None, u'imported': False, u'name': u'Test Ptable', u'changed': False, u'additional_errors': u"Skipping, 'name' filtered out based on 'filter' and 'negate' settings", u'file': u'ptable1.erb', u'type': None, u'id': None}, {u'validation_errors': {}, u'exception': None, u'imported': False, u'name': u'Test Snippet', u'changed': False, u'additional_errors': u"Skipping, 'name' filtered out based on 'filter' and 'negate' settings", u'file': u'snippet1.erb', u'type': None, u'id': None}, {u'validation_errors': {}, u'exception': None, u'imported': False, u'name': u'CoreTest', u'changed': False, u'additional_errors': u"Skipping, 'name' filtered out based on 'filter' and 'negate' settings", u'file': u'core.erb', u'type': None, u'id': None}, {u'validation_errors': {}, u'exception': None, u'imported': False, u'name': u'PluginTest', u'changed': False, u'additional_errors': u"Skipping, 'name' filtered out based on 'filter' and 'negate' settings", u'file': u'plugin.erb', u'type': None, u'id': None}], repo=u'https://github.com/jyejare/foreman_templates.git', branch=None)

TemplateExport Entity:

entities.TemplateExport(repo='/usr/share/foreman_templates/jitu', organization=[1], filter='robo
    ...: ttelo').create()
Out[10]: nailgun.entities.TemplateExport(templates=[{u'exported': True, u'type': u'job_template', u'id': 141, u'name': u'robottelo Ansible Roles - Ansible Default'}, {u'exported': True, u'type': u'ptable', u'id': 142, u'name': u'robottelo Kickstart default thin'}, {u'exported': True, u'type': u'provisioning_template', u'id': 143, u'name': u'robottelo Kickstart default iPXE'}, {u'exported': True, u'type': u'provisioning_template', u'id': 144, u'name': u'robottelo RancherOS provision'}, {u'exported': True, u'type': u'provisioning_template', u'id': 145, u'name': u'robottelo Grubby default'}, {u'exported': True, u'type': u'provisioning_template', u'id': 146, u'name': u'robottelo ansible_provisioning_callback'}, {u'exported': True, u'type': u'provisioning_template', u'id': 147, u'name': u'robottelo AutoYaST default user data'}], error=None, repo=u'/usr/share/foreman_templates/jitu', warning=None, git_user=u'admin', branch=None)
```

**Comment by abalakh** [comment]
>PS: The read() cannot be called on this entities directly, it is just used when create returns data.

This sounded suspicious so i've double checked API docs. According to docs, there're no TemplateImport and TemplateExport entities, but there's Template entity with 2 helpers - import and export accordingly. Creating 2 separate entities with `read()` which shouldn't be called and is overridden with lots of fields ignored doesn't seem right to me.

**Comment by jyejare** [comment]
@abalakh , 
1. Already there are specific entities created in Nailgun to read the entities(those are nothing but JobTemplate, Provisioning Template and Partition Table) you imported.
2. I thought to Create Template entity and then have import and export functions earlier but both import and export returns different output data so thought to have them separate.
3.  Also if I have combine import and export in single Template entity, there will be lot of ignores.

So, What you suggest ?

**Comment by abalakh** [comment]
1. All mentioned entities are actually separate entities - take a look at apidocs, there's separate section per each. Same way there's `Template` entity with 2 paths - import and export, so i think we should do the same
2 and 3. Outputs of helper methods isn't proceeded in any way and returns in a raw dict, so you won't have to ignore anything. And since there's no way to read entity i don't think you will even have to define any fields for it

**Comment by oshtaier** [comment]
Okay, we have nothing like that when you look through whole code base.
We have strict standard for each term or word in nailgun otherwise it will never work for so many years.

`/api/templates/import` is a POST method not a separate piece of logic that we call entity
so
`POST /api/templates/import` and `POST /api/templates/export` should be defined as methods for class `Template`
Yes, we have classes for Provisioning Templates and Config Templates, so it can be not so clear from first glance what `Template` is, but you should not care about that as APIdoc is one and only one source for anyone to understand relations between all and everything.

**Comment by abalakh** [comment]
@jyejare i agree with @oshtaier , we always rely on apidocs and all the entities are created according to apidoc, let's not make an exclusion here, especially with adding some read() methods which can't be called directly and other confusing stuff.

**Comment by jyejare** [comment]
@abalakh , @oshtaier , The new implementation of Template entity - https://github.com/SatelliteQE/nailgun/pull/499/files#diff-3822710a997c3f2f7c3e93e884af69caR6582

When you are ok with that, I will remove the previous one.

**Comment by oshtaier** [comment]
Do these helpers work? Have a chance to test them?

**Comment by jyejare** [comment]
@oshtaier ,

Import:
```
entities.Template().imports(
   ...:     data={
   ...:         'repo':'https://github.com/jyejare/foreman_templates.git',
   ...:         'filter':'robottelo',
   ...:         'prefix':'new__'
   ...:     }
   ...: )
Out[3]: 
{u'message': {u'branch': None,
  u'repo': u'https://github.com/jyejare/foreman_templates.git',
  u'templates': [{u'additional_errors': u"Skipping, 'name' filtered out based on 'filter' and 'negate' settings",
    u'changed': False,
    u'exception': None,
    u'file': u'metadata1.erb',
    u'id': None,
    u'imported': False,
    u'name': u'new__Test Data',
    u'type': None,
    u'validation_errors': {}},
   {u'additional_errors': u"Skipping, 'name' filtered out based on 'filter' and 'negate' settings",
    u'changed': False,
    u'exception': None,
    u'file': u'metadata2.erb',
    u'id': None,
    u'imported': False,
    u'name': u'new__Model-based Test',
    u'type': None,
    u'validation_errors': {}}]}}
```
Export:

```
entities.Template().exports(
   ...:     data={
   ...:         'repo':'/usr/share/foreman_templates/jitu/',
   ...:         'filter':'jenkins',
   ...:         'negate': True,
   ...:         'metadata_export_mode': 'refresh'
   ...:     }
   ...: )
Out[5]: 
{u'message': {u'branch': None,
  u'error': None,
  u'git_user': u'admin',
  u'repo': u'/usr/share/foreman_templates/jitu/',
  u'templates': [{u'exported': True,
    u'id': 1,
    u'name': u'PXELinux global default',
    u'type': u'provisioning_template'},
   {u'exported': True,
    u'id': 2,
    u'name': u'PXEGrub global default',
    u'type': u'provisioning_template'},
   {u'exported': True,
    u'id': 3,
    u'name': u'PXEGrub2 global default',
    u'type': u'provisioning_template'},
   {u'exported': True,
    u'id': 4,
    u'name': u'PXELinux default local boot',
    u'type': u'provisioning_template'},
   {u'exported': True,
    u'id': 5,
    u'name': u'PXEGrub default local boot',
    u'type': u'provisioning_template'},
   {u'exported': True,
    u'id': 6,
    u'name': u'PXEGrub2 default local boot',
    u'type': u'provisioning_template'},
   {u'exported': True,
    u'id': 7,
    u'name': u'PXELinux default memdisk',
    u'type': u'provisioning_template'},
   {u'exported': True,
    u'id': 8,
    u'name': u'PXELinux chain iPXE',
    u'type': u'provisioning_template'},
   {u'exported': True,
    u'id': 9,
    u'name': u'PXELinux chain iPXE UNDI',
    u'type': u'provisioning_template'},
   {u'exported': True,
    u'id': 50,
    u'name': u'WAIK default PXELinux',
    u'type': u'provisioning_template'},
   {u'exported': True,
    u'id': 10,
    u'name': u'Alterator default',
    u'type': u'provisioning_template'},
   {u'exported': True,
    u'id': 11,
    u'name': u'Alterator default finish',
    u'type': u'provisioning_template'},
   {u'exported': True,
    u'id': 12,
    u'name': u'Alterator default PXELinux',
    u'type': u'provisioning_template'},
   {u'exported': True,
    u'id': 13,
    u'name': u'Atomic Kickstart default',
    u'type': u'provisioning_template'},
   {u'exported': True,
    u'id': 14,
    u'name': u'AutoYaST default',
    u'type': u'provisioning_template'}],
 'warning': None}}
```

Proves, it works.

**Comment by jyejare** [comment]
@SatelliteQE/quality-engineers , Can somebody ACK and Merge this on behalf of @abalakh  and @oshtaier  as they are on PTO.

**Comment by sghai** [comment]
Merging the PR for now..
@oshtaier @abalakh : please feel free to comment in case you feel some gap here. Jitendra will fix that later

**Pull request merged**